trieste/acquisition/rule.py

```diff
@@ -1103,6 +1115,380 @@ def state_func(
+        """The location of the search space."""
+        return self._location
+
+    @location.setter
+    def location(self, location: TensorType) -> None:
+        """Set the location of the search space."""
+        self._location = location
+
+    def reinitialise(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> None:
+        """Reinitialise the box."""
+        _, dataset = self.get_single_model_and_dataset(models, datasets)
+
+        self.location = tf.squeeze(self.global_search_space.sample(1), axis=0)
+
+        global_lower = self.global_search_space.lower
+        global_upper = self.global_search_space.upper
+
+        if self._initialised:
+            scale = 0.1
+        else:
+            scale = 0.5
+            self._initialised = True
+
+        self._eps = scale * (global_upper - global_lower) / (5.0 ** (1.0 / global_lower.shape[-1]))
+        self._lower = tf.reduce_max([global_lower, self.location - self._eps], axis=0)
+        self._upper = tf.reduce_min([global_upper, self.location + self._eps], axis=0)
+
+        _, self.y_min = self.get_local_min(dataset)
+
+    def update(
+        self,
+        models: Optional[Mapping[Tag, ProbabilisticModelType]] = None,
+        datasets: Optional[Mapping[Tag, Dataset]] = None,
+    ) -> None:
+        """Update this box, including centre/location, using the given dataset. If the size of the
+        box is less than the minimum size, reinitialise the box."""
+        _, dataset = self.get_single_model_and_dataset(models, datasets)
+
+        if tf.reduce_any(self._eps < self._min_eps):
+            self.reinitialise(models, datasets)
+            return
+
+        _, y_min = self.get_local_min(dataset)
+
+        tr_volume = tf.reduce_prod(self.upper - self.lower)
+        step_is_success = y_min < self.y_min - self._kappa * tr_volume
+        self._eps = self._eps / self._beta if step_is_success else self._eps * self._beta
+
+        self._lower = tf.reduce_max(
+            [self.global_search_space.lower, self.location - self._eps], axis=0
+        )
+        self._upper = tf.reduce_min(
+            [self.global_search_space.upper, self.location + self._eps], axis=0
+        )
+
+        _, self.y_min = self.get_local_min(dataset)
+
+    def get_local_min(self, dataset: Optional[Dataset]) -> Tuple[TensorType, TensorType]:
+        """Calculate the local minimum of the box using the given dataset."""
+        if dataset is None:
+            raise ValueError("""dataset must be provided""")
+
+        in_tr = self.contains(dataset.query_points)
+        in_tr_obs = tf.where(
+            tf.expand_dims(in_tr, axis=-1),
+            dataset.observations,
+            tf.constant(np.inf, dtype=dataset.observations.dtype),
+        )
+        ix = tf.argmin(in_tr_obs)
+        x_min = tf.gather(dataset.query_points, ix)
+        y_min = tf.gather(in_tr_obs, ix)
```

Comment:
I know this was Victor's implementation, but its actually incorrect, it was a hack - say that the batch size is 50, here we check what was the best observation collected in all 50 TRs, while we actually want to get the observation corresponding to the point optimised with a particular TR in the previous iteration

Reply:
Even-though it would start with observations from all 50 TRs, it will only consider the subset of points that are in the current TR. I am assuming this is what you mean by a hack. The downside being that if TRs overlap, the optimised point in the current TR may have been queried by a different region. So are we saying it will be better if we maintained separate datasets for each TR?

Reply:
We plan to integrate TURBO with these classes in a subsequent PR. TURBO would require us to put in place a mechanism to handle local models and datasets. My suggestion is to defer the topic of local datasets to that PR.

Reply:
I agree this is a hack that needs fixing, as we have done it in SD, which is to use the latest query point only that corresponds to this TR. 

I would suggest to write an issue and merge?

Reply:
Issue created.